plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/InteractiveLayeredGraphVisitor.java

```diff
@@ -0,0 +1,453 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Kiel University and others.
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0  
+ *******************************************************************************/
+package org.eclipse.elk.alg.layered;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.elk.alg.layered.options.CycleBreakingStrategy;
+import org.eclipse.elk.alg.layered.options.LayerConstraint;
+import org.eclipse.elk.alg.layered.options.LayeredOptions;
+import org.eclipse.elk.alg.layered.options.LayeringStrategy;
+import org.eclipse.elk.core.math.KVector;
+import org.eclipse.elk.core.options.CoreOptions;
+import org.eclipse.elk.core.options.Direction;
+import org.eclipse.elk.core.util.IGraphElementVisitor;
+import org.eclipse.elk.graph.ElkConnectableShape;
+import org.eclipse.elk.graph.ElkEdge;
+import org.eclipse.elk.graph.ElkGraphElement;
+import org.eclipse.elk.graph.ElkNode;
+import org.eclipse.elk.graph.ElkPort;
+
+/**
+ * Graph visitor which visits only the root node and recursively steps through the graph
+ * to set interactive options and pseudo coordinates for the layered algorithm.
+ * Interactive layout works via two consecutive layout runs.
+ * The first one is a "normal" run. The second one will is configured using this graph visitor.
+ * The graph visitor recursively sets the interactive strategies for all phases and assigns pseudo positions to all
+ * node based on their last position and the desired layer and position in the layer expressed by the
+ * {@code LayerChoiceConstraint} and {@code PositionChoiceConstraint}.
+ * The pseudo position have to represent layers and the ordering in the layer.
+ */
+public class InteractiveLayeredGraphVisitor implements IGraphElementVisitor {
+
+    /**
+     * Since nodes should not overlap and the their width and height may not be final.
+     * Therefore, they are placed with the following spacing, which is hopefully enough to make up for size
+     * changes caused by interactive layout.
+     */
+    public static final int PSEUDO_POSITION_SPACING = Integer.MAX_VALUE;
```

Comment:
Now that I come to think about it, why not iterate over the graph and use the widest node (plus a small number) as the pseudo position spacing?

Reply:
The problem is I do not know how wide the widest node is and this could change by something which is related to the area of the biggest node.
If I have a node that is 100 wide and 100 height before interactive layout. Than interactive layout might force me to draw all 10000 1 unit width nodes next to each other leaving a width of 10000 (no spacing). I think if might get into problems doing this because I have to consider all descendants of a node to make sure it cannot become too big. My example does of course only work for nodes bigger than 1. Using the area as additional width might however be a good approximation and is probably better than unsing `MAX_INT`.

Reply:
Ah, because you need to configure the layout graph before the nodes have their new sizes set, that makes sense. This discussion is an indication of what needs to be documented here. 🙂